client/server (or server/repository) setup with SSH keys exchange

anarcat opened the pull request:
WIP.

Depends on #4

anarcat committed:
start configuring client/server (or server/repository) setup

anarcat committed:
SSH keys support

anarcat committed:
make one user per postgresql server

anarcat committed:
disable SSH code on repository correctly if turned off

anarcat committed:
switch to ssh_keygen module

anarcat committed:
ssh_authorized_key doesn't have a comment field, of course

anarcat committed:
fix user creation code

We were missing an ensure, dependencies, and in general weren't clear
on usernames.

anarcat committed:
create missing directories in config_file

anarcat committed:
document more TODO items

anarcat committed:
fix permissions on ssh_authorized_keys

anarcat committed:
manage config files at arbitrary depths

Also purge unmanaged configuration files, but offer a knob to turn
that off.

anarcat committed:
create stanza and isolate hosts

anarcat committed:
manage directories properly

anarcat committed:
fix duplicate declaration in ini settings

anarcat committed:
add timers to perform regular full/diff backups

anarcat committed:
fix stanza creation

anarcat committed:
more docs: explain archive_command

anarcat commented:
so. small progress update on this: this is mostly functional. i can setup new servers automatically with a central server, each node is isolated from one another with sandbox users, all configured over SSH automatically, it's pretty neat!

i still need to figure out how to configure expiry properly, and i have a bunch of hardcoded paths in there, so it's not yet for public consumption (still draft!) but I wanted to let people know here that it's pretty close to completion.

i'm not sure i'll have the stamina to finish this for that, as there's a lot of glue code that's specific to our stuff, but if people are interested, that will help me get motivated! :)

in any case, we're likely going to be using this in production going forward. nearing the end of our alpha phase here, we just need to test restores to see if this all works like we want, and then we're off the races to drop our old legacy backup system in favor of this!

anarcat committed:
move high priority TODO items to issue https://gitlab.torproject.org/tpo/tpa/team/-/issues/40950

anarcat committed:
properly manage non-default base directory

anarcat committed:
handle connection failures when bootstrapping stanzas properly

ubellavance commented:
I am not too familiar with drafts or WIP PRs.  If I don't see your final PR, feel free to contact me by writing a comment.

anarcat commented:
I will certainly let you know of progress here, don't worry!

At this stage, this is working in production for us. It requires some
module dependencies that are not correctly specified in the module's
metadata, and there's some hardcoded paths for our peculiar SSH
authorized_keys setup, but it otherwise works for us!

I'll take a pause on developing this until at least January 2025. I'm
hoping to resume work on this in 2025, but I'm not sure I'll have time
to complete this to make it publicly usable by others.

I'll keep maintaining our branch however, that's for sure, and I'm happy
to collaborate on or review work from others interested in this feature
set.

And of course, ideally, I'll just finish this soon. :)

Stay tuned, and don't hesitate to ping me if you want news!

anarcat committed:
make backup schedules configurable

anarcat committed:
pgbackrest: persist long-running calendar jobs across reboots

This should make those timers more reliable as they'll keep track of
multi-days (e.g. full, monthly) backup schedules better.

anarcat committed:
extend backup schedule to 30 days

We switch to one full per month, and use the "incremental" feature to
stagger backups during the month.

anarcat committed:
fix permissions on /var/log

This should fix that warning:

    WARN: unable to open log file '/var/log/pgbackrest/rude.torproject.org/rude.torproject.org-expire.log': Permission denied
          NOTE: process will continue without log file.

anarcat committed:
use simpler Service resource to enable units

This removes the need to actually write a symlink, and instead uses
the normal systemd provider mechanisms to activate the timer.

ubellavance commented:
@anarcat FWIW, I was told by one of `pgbackrest`'s developpers that it would be preferable to implement this using TLS instead of SSH. He added "We haven’t deprecated SSH but we consider TLS to be the future — not least because it is much faster than SSH."

anarcat commented:
> @anarcat FWIW, I was told by one of `pgbackrest`'s developpers that it would be preferable to implement this using TLS instead of SSH. He added "We haven’t deprecated SSH but we consider TLS to be the future — not least because it is much faster than SSH."

Yeah, I understand. I've talked with some pgbackrest people about this,
and I am not a fan of running another TLS server on all my pg servers
right now. If we do have performance issues, I'll reroll this another
way, but for now that's how it's deployed...

Note that, anyways, the way it's done now is just not portable outside
our infra, the main reason why it's still marked as draft. That's
because we rely on SSH stuff that's not specified in the deps. It's all
kind of a mess, see also our internal issue about this:

https://gitlab.torproject.org/tpo/tpa/team/-/issues/41890

... so, in a sense, maybe TLS would be a solution there... But I'm still
struggling to just finish this project and retire legacy, I'm not sure I
can just in another refactoring just yet. I barely have things working
right now (especially schedules and retentions were tricky, still in
testing...)

ubellavance commented:
No worries, there is no rush.

dwsteele commented:
> I've talked with some pgbackrest people about this

I'm curious -- when/where did we talk? There aren't that many pgBackRest people...

Don't get me wrong, I don't think there is a problem with implementing this on SSH. We have not deprecated SSH and don't plan to. We just recommend TLS because it is faster and does not have the problem of motd, etc. spewing garbage into the connection.

anarcat commented:
>> I've talked with some pgbackrest people about this
>
> I'm curious -- when/where did we talk? There aren't that many pgBackRest people...

Maybe we didn't talk and I got confused!

What did happen for sure as I asked a lot of pgbackrest newbie questions
on IRC (irc.libera.chat/#postgresql)...

> Don't get me wrong, I don't think there is a problem with implementing this on SSH. We have not deprecated SSH and don't plan to. We just recommend TLS because it is faster and does not have the problem of motd, etc. spewing garbage into the connection.

Right. I think I understand the tradeoffs pretty well. My priority at
this point is getting rid of our legacy backup system. If you want all
the gory details, it's at:

https://gitlab.torproject.org/tpo/tpa/team/-/issues/40950

but i've added back "TLS" to our roadmap because of the multiple
comments here. it *might* be easier to manage in Puppet as well,
although we'd need to manage TLS certs instead of SSH, which is kind of
a whole ball game...

Right now we have a private CA for that stuff, but it's kind of out of
band, so we'd need to hook *something* up in this module for it to make
any sort of sense (I don't want to assume cleartext is Fine).

ubellavance commented:
FWIW, I manage TLS certs for MariaDB using Puppet (eyaml) and it's not that complex.

anarcat commented:
> FWIW, I manage TLS certs for MariaDB using Puppet (eyaml) and it's not that complex.

you'll be absolutely delighted to know that we manage TLS certs with puppet, certbot, makefiles, and god knows what else I forgot: https://gitlab.torproject.org/tpo/tpa/team/-/wikis/howto/tls/#certificate-authorities-in-use-at-tor

there's only so much legacy i can fix per day. ;)

dwsteele commented:
> What did happen for sure as I asked a lot of pgbackrest newbie questions on IRC (irc.libera.chat/#postgresql)...

Well, plenty of pgbackrest users out there. I was thinking devs but that's just my dev-centric view. The users are the point.

> although we'd need to manage TLS certs instead of SSH, which is kind of
a whole ball game...

No question TLS is harder to manage and getting access to the CA is a big part of that.

dwsteele commented:
> there's only so much legacy i can fix per day. ;)

A truer thing was never said!

dwsteele commented:
> If you want all the gory details

I read through the thread and it was very interesting to see how the whole thing evolved over time.

> one major caveat is it seems more "open" than barman by default, with every client having access to every backup, so we'll need to nail this down. we'll also need to test restores and setup the exporter.

You'll be glad (maybe) to know that this is improved in the TLS implementation. Each client can be restricted to a specific stanza list. We have also limited this with SSH in the past by limiting the `--stanza` option that can be used in execs but it is a bit ugly and it has been so long I'm not even sure it would work with recent implementations.

BTW -- I'm a huge fan of the Tor project. It is such an important project so I'm happy you are using pgBackRest!

anarcat commented:
On 2025-02-04 14:14:43, David Steele wrote:
>> If you want all the gory details
>
> I read through the thread and it was very interesting to see how the whole thing evolved over time.

Oh my, thanks for reviewing this, that's awesome!

>> one major caveat is it seems more "open" than barman by default, with every client having access to every backup, so we'll need to nail this down. we'll also need to test restores and setup the exporter.
>
> You'll be glad (maybe) to know that this is improved in the TLS implementation. Each client can be restricted to a specific stanza list. We have also limited this with SSH in the past by limiting the `--stanza` option that can be used in execs but it is a bit ugly and it has been so long I'm not even sure it would work with recent implementations.

Oh that is *really* interesting. That's one thing I've struggled with a
*lot* in the current setup: I had to create users per stanza to isolate
each server, it's a mess... That could be much simpler indeed.

> BTW -- I'm a huge fan of the Tor project. It is such and important project and I'm happy that you are using pgBackRest!

Aw, that's sweet thanks!

anarcat committed:
fix warning with null schedules

Puppet was still trying to start the incr timers:

Error: Systemd start for pgbackrest-backup-incr@weather-01.timer failed!
journalctl log for pgbackrest-backup-incr@weather-01.timer:
Feb 12 21:43:14 backup-storage-01 systemd[1]: pgbackrest-backup-incr@weather-01.timer: Timer unit lacks value setting. Refusing.

... because the incr thing *is* in the key.

anarcat committed:
pgbackrest: fix schedules (tpo/tpa/team#42043)

The problem with the previous "monthly" and "weekly" configurations is
that they were splaying the results from *after* the calendar
date. This *eventually* works, it converges over the planned
frequency, but when systemd restarts or the server reboots, things go
out of whack and there will be additional delays.

I consider this to be a bug in systemd, see:

https://github.com/systemd/systemd/issues/21166

So instead we pick a random day of the month or of the week in
Puppet. We also shift the definition to the client-side, because
otherwise the randomization happens on the server, and only on the
server, which defeats the purpose as all instances of the template
will run at the same time.

anarcat committed:
pgbackrest: recrate lock directories on boot

Those files are purged automatically on reboot because they are
actually in /run/lock. Our previous approach worked but only after
puppet ran, after boot, which meant that for a couple of hours after
reboot, we would have an outage.

anarcat committed:
retry backups on failure

We had a situation where a backup had failed, presumably because it
was performed while the host was rebooting. It seems fair to try to
reboot those correctly.

We introduce exponential backoff here, but that's supported only in
systemd 254 (trixie and later). It's ignored by earlier versions
though, so that's safe to include here.

We limit the number of restarts per day to 10 as well.

anarcat committed:
pass cluster version down into the stanza

anarcat committed:
fix crash without exported resources

This was working in production for us because we already had config
object that were creating the parent directory, this crashes
otherwise with:

Error: /Stage[main]/Pgbackrest::Config/File[/etc/pgbackrest/conf.d]/ensure: change from 'absent' to 'directory' failed: Cannot create /etc/pgbackrest/conf.d; parent directory /etc/pgbackrest does not exist

anarcat committed:
fix ssh_keygen home directory

By default, ssh_keygen does *not* lookup the user's home before
creating the key: it just assumes /home (or /root), which fails in
this case.

Instead of leaving this hardcoded, make it customizable as well so we
can support other platforms and layouts, and pass it down into the
stanza.

Also order the user key creation correctly.
